### PR TITLE
Add GnuPG dynamic application built from source

### PR DESCRIPTION
**Pull request by andreistan26**
#47
Add GnuPG dynamic application built from source. Added the `build.sh` file for building the app.

### CODE REVIEW

**Review by RaduNichita — CHANGES_REQUESTED** [review]
You don't need to include the libraries or other binary file, only the `build.sh` script and a `README.md` file in which you explain how to run the application on both Linux and Unikraft.

You can see an example [here](https://github.com/unikraft/dynamic-apps/pull/53)

File: gnupg_source/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# GnuPG
+
+In order to build `GnuPG` you have to run, from the **gnupg_source** directory
+```
```
Comment (StefanJum) [inline gnupg_source/README.md:4]:
```suggestion
```console
```

File: gnupg_source/README.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# GnuPG
+
+In order to build `GnuPG` you have to run, from the **gnupg_source** directory
```
Comment (StefanJum) [inline gnupg_source/README.md:3]:
```suggestion
In order to build `GnuPG` you have to run, from the **gnupg_source** directory

```

File: gnupg_source/README.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+```
+./build.sh
+```
+
+The script will download GnuPG and it's dependecies, build them and use
+the `extract.sh` tool to generate the `lib` `lib64` and `bin` directories.
```
Comment (StefanJum) [inline gnupg_source/README.md:9]:
```suggestion
The script will download GnuPG and it's dependecies, build them and use the `extract.sh` tool to generate the `lib` `lib64` and `bin` directories.
```
Use one line per sentence and one sentence per line.

File: gnupg_source/build.sh
Comment (StefanJum) [inline gnupg_source/build.sh:1]: Run the script through [shellcheck](https://www.shellcheck.net/).

File: gnupg_source/build.sh [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+LIBGPG_ERROR=libgpg-error-1.47
+LIBGC_CRYPT=libgcrypt-1.10.2
+LIBASSUAN=libassuan-2.5.6
+LIBKSBA=libksba-1.6.4
+LIBNPTH=npth-1.6
+
+cp ../extract.sh .
+
+if [ ! -f "extract.sh" ]; then
+    echo "You need extract.sh in order to build!"
```
Comment (StefanJum) [inline gnupg_source/build.sh:13]:
```suggestion
    echo "You need extract.sh in order to build!" 1>&2
```

File: gnupg_source/build.sh [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    exit 1
+fi
+
+# Downloading dependencies
+
+echo "Donwloading dependencies ..."
+
+## libgpg-erro
+wget -q https://gnupg.org/ftp/gcrypt/gpgrt/$LIBGPG_ERROR.tar.bz2
```
Comment (StefanJum) [inline gnupg_source/build.sh:22]:
```suggestion
wget -q "https://gnupg.org/ftp/gcrypt/gpgrt/$LIBGPG_ERROR.tar.bz2"
```

File: gnupg_source/build.sh [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+echo ""
+
+tar xvf $LIBGPG_ERROR.tar.bz2 && rm -f $LIBGPG_ERROR.tar.bz2
```
Comment (StefanJum) [inline gnupg_source/build.sh:26]:
```suggestion
tar xvf "$LIBGPG_ERROR.tar.bz2" && rm -f "$LIBGPG_ERROR.tar.bz2"
```

File: gnupg_source/build.sh [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+
+cd $LIBGPG_ERROR
```
Comment (StefanJum) [inline gnupg_source/build.sh:29]:
```suggestion
pushd "$LIBGPG_ERROR" &> /dev/null || exit 1
```

File: gnupg_source/build.sh [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+./configure && make -j $(nproc) && sudo make install -j $(nproc)
+
+if [ $? -ne 0 ]; then
+    echo "Building $LIBGPG_ERROR failed!"
+    exit
```
Comment (StefanJum) [inline gnupg_source/build.sh:34]:
```suggestion
    exit 1
```

File: gnupg_source/build.sh [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+echo ""
+
+tar xvf $LIBGPG_ERROR.tar.bz2 && rm -f $LIBGPG_ERROR.tar.bz2
+
+
+cd $LIBGPG_ERROR
+./configure && make -j $(nproc) && sudo make install -j $(nproc)
+
+if [ $? -ne 0 ]; then
+    echo "Building $LIBGPG_ERROR failed!"
```
Comment (StefanJum) [inline gnupg_source/build.sh:33]:
```suggestion
    echo "Building $LIBGPG_ERROR failed!" 1>&2
```

File: gnupg_source/build.sh [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    exit
+fi
+
+cd ..
```
Comment (StefanJum) [inline gnupg_source/build.sh:37]:
```suggestion
popd &> /dev/null || exit 1
```

File: gnupg_source/build.sh [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+
+## libgcrypt
+wget -q https://gnupg.org/ftp/gcrypt/libgcrypt/$LIBGC_CRYPT.tar.bz2
```
Comment (StefanJum) [inline gnupg_source/build.sh:40]:
```suggestion
wget -q "https://gnupg.org/ftp/gcrypt/libgcrypt/$LIBGC_CRYPT.tar.bz2"
```

File: gnupg_source/build.sh [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+echo ""
+
+tar xvf $LIBGC_CRYPT.tar.bz2 && rm -f $LIBGC_CRYPT.tar.bz2
```
Comment (StefanJum) [inline gnupg_source/build.sh:44]:
```suggestion
tar xvf "$LIBGC_CRYPT.tar.bz2" && rm -f "$LIBGC_CRYPT.tar.bz2"
```

File: gnupg_source/build.sh [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+cd $LIBGC_CRYPT
```
Comment (StefanJum) [inline gnupg_source/build.sh:46]:
```suggestion
pushd "$LIBGC_CRYPT" &> /dev/null || exit 1
```

File: gnupg_source/build.sh [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+./configure && make -j $(nproc) && sudo make install -j $(nproc)
+
+if [ $? -ne 0 ]; then
+    echo "Building $LIBGC_CRYPT failed!"
+    exit
+fi
+
+cd ..
```
Comment (StefanJum) [inline gnupg_source/build.sh:54]:
```suggestion
popd &> /dev/null || exit 1
```

File: gnupg_source/build.sh [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+
+## libassuan
+wget -q https://gnupg.org/ftp/gcrypt/libassuan/$LIBASSUAN.tar.bz2
+
+echo ""
+
+tar xvf $LIBASSUAN.tar.bz2 && rm -f $LIBASSUAN.tar.bz2
+
+cd $LIBASSUAN
```
Comment (StefanJum) [inline gnupg_source/build.sh:63]:
```suggestion
pushd "$LIBASSUAN" &> /dev/null || exit 1
```

File: gnupg_source/build.sh [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+./configure && make -j $(nproc) && sudo make install -j $(nproc)
+
+if [ $? -ne 0 ]; then
+    echo "Building $LIBASSUAN failed!"
+    exit
+fi
+
+cd ..
```
Comment (StefanJum) [inline gnupg_source/build.sh:71]:
```suggestion
popd &> /dev/null || exit 1
```

File: gnupg_source/build.sh [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+
+if [ $? -ne 0 ]; then
+    echo "Building $LIBGC_CRYPT failed!"
+    exit
+fi
+
+cd ..
+
+## libassuan
+wget -q https://gnupg.org/ftp/gcrypt/libassuan/$LIBASSUAN.tar.bz2
```
Comment (StefanJum) [inline gnupg_source/build.sh:57]:
```suggestion
wget -q "https://gnupg.org/ftp/gcrypt/libassuan/$LIBASSUAN.tar.bz2"
```

File: gnupg_source/build.sh [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+
+echo ""
+
+tar xvf $LIBASSUAN.tar.bz2 && rm -f $LIBASSUAN.tar.bz2
```
Comment (StefanJum) [inline gnupg_source/build.sh:61]:
```suggestion
tar xvf "$LIBASSUAN.tar.bz2" && rm -f "$LIBASSUAN.tar.bz2"
```

File: gnupg_source/build.sh [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+
+if [ $? -ne 0 ]; then
+    echo "Building $LIBASSUAN failed!"
+    exit
+fi
+
+cd ..
+
+## libksba
+wget -q https://gnupg.org/ftp/gcrypt/libksba/libksba-1.6.4.tar.bz2
```
Comment (StefanJum) [inline gnupg_source/build.sh:74]:
```suggestion
wget -q "https://gnupg.org/ftp/gcrypt/libksba/libksba-1.6.4.tar.bz2"
```

File: gnupg_source/build.sh [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+
+echo ""
+tar xvf $LIBKSBA.tar.bz2 && rm -f $LIBKSBA.tar.bz2
```
Comment (StefanJum) [inline gnupg_source/build.sh:77]:
```suggestion
tar xvf "$LIBKSBA.tar.bz2" && rm -f "$LIBKSBA.tar.bz2"
```

File: gnupg_source/build.sh [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+
+cd $LIBKSBA
```
Comment (StefanJum) [inline gnupg_source/build.sh:79]:
```suggestion
pushd "$LIBKSBA" &> /dev/null || exit 1
```

File: gnupg_source/build.sh [hunk 22]
```diff
@@ -0,0 +1,151 @@
...
+./configure && make -j $(nproc) && sudo make install -j $(nproc)
+
+if [ $? -ne 0 ]; then
+    echo "Building $LIBKSBA failed!"
+    exit
+fi
+
+cd ..
+
+
```
Comment (StefanJum) [inline gnupg_source/build.sh:89]:
```suggestion

```

File: gnupg_source/build.sh [hunk 23]
```diff
@@ -0,0 +1,151 @@
...
+cd $LIBKSBA
+./configure && make -j $(nproc) && sudo make install -j $(nproc)
+
+if [ $? -ne 0 ]; then
+    echo "Building $LIBKSBA failed!"
+    exit
+fi
+
+cd ..
```
Comment (StefanJum) [inline gnupg_source/build.sh:87]:
```suggestion
popd &> /dev/null || exit 1
```

File: gnupg_source/build.sh [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+
+
+## libnpth
+wget -q https://gnupg.org/ftp/gcrypt/npth/npth-1.6.tar.bz2
```
Comment (StefanJum) [inline gnupg_source/build.sh:91]:
```suggestion
wget -q "https://gnupg.org/ftp/gcrypt/npth/npth-1.6.tar.bz2"
```

File: gnupg_source/build.sh [hunk 25]
```diff
@@ -0,0 +1,151 @@
...
+
+echo ""
+tar xvf $LIBNPTH.tar.bz2 && rm -f $LIBNPTH.tar.bz2
```
Comment (StefanJum) [inline gnupg_source/build.sh:94]:
```suggestion
tar xvf "$LIBNPTH.tar.bz2" && rm -f "$LIBNPTH.tar.bz2"
```

File: gnupg_source/build.sh [hunk 26]
```diff
@@ -0,0 +1,151 @@
...
+
+cd $LIBNPTH
```
Comment (StefanJum) [inline gnupg_source/build.sh:96]:
```suggestion
pushd "$LIBNPTH" &> /dev/null || exit 1
```

File: gnupg_source/build.sh [hunk 27]
```diff
@@ -0,0 +1,151 @@
...
+./configure && make -j $(nproc) && sudo make install -j $(nproc)
+
+if [ $? -ne 0 ]; then
+    echo "Building $LIBNPTH failed!"
+    exit
+fi
+
+cd ..
```
Comment (StefanJum) [inline gnupg_source/build.sh:104]:
```suggestion
popd &> /dev/null || exit 1
```

File: gnupg_source/build.sh [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+
+wget -q https://gnupg.org/ftp/gcrypt/gnupg/gnupg-2.4.3.tar.bz2
+
+echo ""
+
+echo "Extracting gnupg ..."
+
+tar xvf $LIBGNUPG.tar.bz2 && rm -rf $LIBGNUPG.tar.bz2
+
+cd $LIBGNUPG
```
Comment (StefanJum) [inline gnupg_source/build.sh:119]:
```suggestion
pushd "$LIBGNUPG" &> /dev/null || exit 1
```

File: gnupg_source/build.sh [hunk 29]
```diff
@@ -0,0 +1,151 @@
...
+
+./configure && make -j $(nproc) && sudo make install -j $(nproc)
+
+if [ $? -ne 0 ]; then
+    echo "Building $LIBGNUPG failed!"
+    exit
+fi
+
+cd ..
```
Comment (StefanJum) [inline gnupg_source/build.sh:128]:
```suggestion
popd &> /dev/null || exit 1
```

File: gnupg_source/build.sh [hunk 30]
```diff
@@ -0,0 +1,151 @@
...
+
+# Cleanup of dependencies
+rm -rf $LIBNPTH
+rm -rf $LIBKSBA
+rm -rf $LIBASSUAN
+rm -rf $LIBGPG_ERROR
+rm -rf $LIBGC_CRYPT
```
Comment (StefanJum) [inline gnupg_source/build.sh:135]:
```suggestion
rm -rf "$LIBNPTH"
rm -rf "$LIBKSBA"
rm -rf "$LIBASSUAN"
rm -rf "$LIBGPG_ERROR"
rm -rf "$LIBGC_CRYPT"
```

File: gnupg_source/build.sh [hunk 31]
```diff
@@ -0,0 +1,151 @@
...
+
+# Extract
+
+declare -a binaries=("g10/gpg" "agent/gpg-agent" "tools/gpg-card" "tools/gpgconf" 
+    "tools/gpgtar" "g10/gpgv" "sm/gpgsm")
+
+mkdir bin
+
+for binary in "${binaries[@]}"; do
+    cp $LIBGNUPG/$binary bin
```
Comment (StefanJum) [inline gnupg_source/build.sh:145]:
```suggestion
    cp "$LIBGNUPG/$binary" bin
```

File: gnupg_source/build.sh [hunk 32]
```diff
@@ -0,0 +1,151 @@
...
+    ./extract.sh $LIBGNUPG/$binary
```
Comment (StefanJum) [inline gnupg_source/build.sh:146]:
```suggestion
    ./extract.sh "$LIBGNUPG/$binary"
```

**Review by StefanJum — CHANGES_REQUESTED** [review]
Thanks @andreistan26, see the comments.
Besides those, it seems like there is a lot of similar code for building different libraries, see if it makes sense to have a bash function that does that (receives an url and a name as parameters and does the job).

**Comment by andreistan26** [comment]
Thanks for all the tips. I've [refactored](https://github.com/unikraft/dynamic-apps/pull/57/commits/0371ae8f398a2a1f8b595b314d01e08147b3c04d?short_path=9e04612#diff-9e04612a2229595cfc780aa59b9fd5662e5540a82d99322752cc66d3a9cea1bc) both the `README` and the `build.sh`. They are ready for another review.

File: gnupg_source/README.md [hunk 33]
```diff
@@ -0,0 +1,29 @@
+# GnuPG as Dynamic Build
+
+This directory stores the `build.sh` file, used for building [GnuPG]("https://gnupg.org/"), also known as `gpg`, a command line tool for
+encrypting and singing your data. All files are stored in their expected
+location in the Linux filesytem hierarchy. The aim is to run the
+executable in any Linux ELF-compatible environment: native Linux, chroot/jail, Docker, Unikraft.
```
Comment (StefanJum) [inline gnupg_source/README.md:6]:
```suggestion
This directory stores the `build.sh` file, used for building [GnuPG]("https://gnupg.org/"), also known as `gpg`, a command line tool for encrypting and singing your data.
All files are stored in their expected location in the Linux filesytem hierarchy.
The aim is to run the executable in any Linux ELF-compatible environment: native Linux, chroot/jail, Docker, Unikraft.
```
Use one line per sentence and one sentence per line

File: gnupg_source/README.md [hunk 34]
```diff
@@ -0,0 +1,29 @@
...
+
+For more about the usage of `gpg` visit the [official documentation](https://www.gnupg.org/documentation/)
```
Comment (StefanJum) [inline gnupg_source/README.md:8]:
```suggestion
For more about the usage of `gpg` visit the [official documentation](https://www.gnupg.org/documentation/).
```

File: gnupg_source/README.md [hunk 35]
```diff
@@ -0,0 +1,29 @@
...
+
+## How to Build
+
+In order to build `GnuPG` you have to run, from inside the **gnupg_source** 
+directory
```
Comment (StefanJum) [inline gnupg_source/README.md:13]:
```suggestion
In order to build `GnuPG` you have to run, from inside the **gnupg_source**  directory.
```
Use one line per sentence and one sentence per line

File: gnupg_source/README.md [hunk 36]
```diff
@@ -0,0 +1,29 @@
...
+
+```console
+./build.sh
+```
+
+The script will download GnuPG and it's dependecies, build them and use
+the `extract.sh` tool to generate the `lib` `lib64` and `bin` directories.
```
Comment (StefanJum) [inline gnupg_source/README.md:20]:
```suggestion
The script will download GnuPG and it's dependecies, build them and use the `extract.sh` tool to generate the `lib` `lib64` and `bin` directories.
```

File: gnupg_source/README.md [hunk 37]
```diff
@@ -0,0 +1,29 @@
...
+
+## How to Run on Unikraft
+
+To run the binary ELF on Unikraft, run the `run.sh` script from the 
+[`run-app-elfloader` repository](https://github.com/unikraft/run-app-elfloader):
```
Comment (StefanJum) [inline gnupg_source/README.md:25]:
```suggestion
To run the binary ELF on Unikraft, run the `run.sh` script from the  [`run-app-elfloader` repository](https://github.com/unikraft/run-app-elfloader):
```

File: gnupg_source/build.sh [hunk 38]
```diff
@@ -0,0 +1,81 @@
...
+echo "Donwloading dependencies ..."
+
+# Building dependencies
+
+get_and_build $LIBGPG_ERROR $LIBGPG_ERROR_URL
+get_and_build $LIBGC_CRYPT $LIBGC_CRYPT_URL
+get_and_build $LIBASSUAN $LIBASSUAN_URL
+get_and_build $LIBKSBA $LIBKSBA_URL
+get_and_build $LIBNPTH $LIBNPTH_URL
+get_and_build $LIBGNUPG $LIBGNUPG_URL
```
Comment (StefanJum) [inline gnupg_source/build.sh:57]:
```suggestion
get_and_build "$LIBGPG_ERROR" "$LIBGPG_ERROR_URL"
get_and_build "$LIBGC_CRYPT" "$LIBGC_CRYPT_URL"
get_and_build "$LIBASSUAN" "$LIBASSUAN_URL"
get_and_build "$LIBKSBA" "$LIBKSBA_URL"
get_and_build "$LIBNPTH" "$LIBNPTH_URL"
get_and_build "$LIBGNUPG" "$LIBGNUPG_URL"
```

**Review by StefanJum — CHANGES_REQUESTED** [review]
Looks a lot better @andreistan26, few more comments, hopefully the last ones.

**Comment by andreistan26** [comment]
Should be fine now.

**Review by StefanJum — CHANGES_REQUESTED** [review]
@andreistan26 I get this error when running `./build.sh`

```
gcc -DWITHOUT_NPTH=1  -std=gnu99 -I/usr/local/include  -I/usr/include/p11-kit-1 -I/usr/local/include -I/usr/local/include -I/usr/local/include -Wall -Wno-format-zero-length -Wno-pointer-sign -Wpointer-arith -g -O2   -o t-http-basic t_http_basic-t-support.o t_http_basic-dns.o t_http_basic-t-http-basic.o t_http_basic-http.o t_http_basic-dns-stuff.o t_http_basic-http-common.o ../common/libcommon.a -L/usr/local/lib -lassuan  -L/usr/local/lib -lgcrypt -L/usr/local/lib -lgpg-error -lgnutls  -lresolv     -L/usr/local/lib -lksba -lgpg-error -L/usr/local/lib -lgcrypt -L/usr/local/lib -lgpg-error -lgnutls -lresolv  
/usr/bin/ld: server.o: in function `cmd_ad_query':
/home/stefan/projects/unikraft/usoc23/session-04/dynamic-apps/gnupg_source/gnupg-2.4.3/dirmngr/server.c:2779: undefined reference to `ks_ldap_help_variables'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:937: dirmngr] Error 1
```

**Comment by andreistan26** [comment]
That is weird, the undefined reference is not even an external dependency. We could try to add  `autoreconf -i` before running `./configure` for the files, maybe that can help. Also I've found the build process (gpg's) to be a little non deterministic, did you try to rerun the build process?

File: gnupg_source/build.sh [hunk 39]
```diff
@@ -0,0 +1,82 @@
...
+
+    echo ""
+
+    tar xvf "$LIB_NAME.tar.bz2" && rm -f "$LIB_NAME.tar.bz2"
+
+    pushd "$LIB_NAME" &> /dev/null || exit 1
+
+    ./autoreconf -i
+    ./configure && make -j "$(nproc)" && sudo make install -j "$(nproc)"
+
```
Comment (StefanJum) [inline gnupg_source/build.sh:17]: There are some trailing whitespaces here

**Review by StefanJum — COMMENTED** [review]
See the one comment.
I still get the undefined reference error. Are you sure you did not install anything `ldap`-related besides running the script?
@RaduNichita @eduardvintila maybe you can try this too and see if it works for you.

File: gnupg_source/build.sh (hunk 39 above)
Comment (andreistan26) [reply gnupg_source/build.sh:17]: Fixing those

File: gnupg_source/README.md [hunk 40]
```diff
@@ -0,0 +1,25 @@
...
+
+## How to Build
+
+In order to build `GnuPG` you have to run, from inside the **gnupg_source** directory
+
+```console
+./build.sh
+```
+
+The script will download GnuPG and it's dependecies, build them and use the `extract.sh` tool to generate the `lib` `lib64` and `bin` directories.
```
Comment (RaduNichita) [inline gnupg_source/README.md:17]:
```suggestion
The script will download GnuPG and it's dependencies, build them and use the `extract.sh` tool to generate the `lib` `lib64` and `bin` directories.
```

File: gnupg_source/build.sh [hunk 41]
```diff
@@ -0,0 +1,82 @@
...
+cp ../extract.sh .
+
+if [ ! -f "extract.sh" ]; then
+    echo "You need extract.sh in order to build!" 1>&2
+    exit 1
+fi
+
+# Downloading dependencies
+
+echo "Donwloading dependencies ..."
```
Comment (RaduNichita) [inline gnupg_source/build.sh:49]:
```suggestion
echo "Downloading dependencies ..."
```

File: gnupg_source/build.sh [hunk 42]
```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
+
+get_and_build() {
+    LIB_NAME=$1
+    LIB_URL=$2
+
+    wget -q "$LIB_URL"
+
+    echo ""
```
Comment (RaduNichita) [inline gnupg_source/build.sh:9]:
```suggestion

```

File: gnupg_source/build.sh [hunk 43]
```diff
@@ -0,0 +1,82 @@
...
+
+    if [ $? -ne 0 ]; then
+        echo "Building $LIB_NAME failed!" 1>&2
+        exit 1
+    fi
+
+    popd &> /dev/null || exit 1
+}
+
+LIBGNUPG="gnupg-2.4.3"
```
Comment (RaduNichita) [inline gnupg_source/build.sh:26]: I suggest that you put those at the beginning of the file.

File: gnupg_source/build.sh [hunk 44]
```diff
@@ -0,0 +1,82 @@
...
+    "tools/gpgtar" "g10/gpgv" "sm/gpgsm")
+
+mkdir bin
+
+for binary in "${binaries[@]}"; do
+    cp "$LIBGNUPG/$binary" bin
+    ./extract.sh "$LIBGNUPG/$binary"
+done
+
+mv usr/local/lib/libgpg-error.so.0 lib/x86_64-linux-gnu
```
Comment (RaduNichita) [inline gnupg_source/build.sh:72]:
```suggestion
mv ./usr/local/lib/libgpg-error.so.0 ./lib/x86_64-linux-gnu
```

File: gnupg_source/README.md [hunk 45]
```diff
@@ -0,0 +1,25 @@
+# GnuPG as Dynamic Build
+
+This directory stores the `build.sh` file, used for building [GnuPG]("https://gnupg.org/"), also known as `gpg`, a command line tool for encrypting and singing your data.
+All files are stored in their expected location in the Linux filesytem hierarchy. 
+The aim is to run the executable in any Linux ELF-compatible environment: native Linux, chroot/jail, Docker, Unikraft.
+
+For more about the usage of `gpg` visit the [official documentation](https://www.gnupg.org/documentation/).
+
+## How to Build
```
Comment (RaduNichita) [inline gnupg_source/README.md:9]:
```suggestion
## Build
```

File: gnupg_source/README.md [hunk 46]
```diff
@@ -0,0 +1,25 @@
...
+
+In order to build `GnuPG` you have to run, from inside the **gnupg_source** directory
+
+```console
+./build.sh
+```
+
+The script will download GnuPG and it's dependecies, build them and use the `extract.sh` tool to generate the `lib` `lib64` and `bin` directories.
+
+## How to Run on Unikraft
```
Comment (RaduNichita) [inline gnupg_source/README.md:19]:
```suggestion
## Run
```

File: gnupg_source/README.md (hunk 40 above)
Comment (RaduNichita) [inline gnupg_source/README.md:17]:
```suggestion
The script will download the GnuPG, install the required dependencies and build the sources.
```

**Review by RaduNichita — CHANGES_REQUESTED** [review]
It's looking better now, @andreistan26, thanks for the effort. I have left another round of feedback.

Please also rename the folder to be `gnupg_src`.

**Comment by RaduNichita** [comment]
Furthermore, modify the commit message so the signoff line would contain your real name, not your Github Id. 

Something line:

```console
Signed-off-by:  Andrei Stan <andreistan2003@gmail.com>
```

**Comment by andreistan26** [comment]
@RaduNichita have you tried running `build.sh` and if so, did you had any issues with it?

**Review by StefanJum — APPROVED** [review]
All good, thank you @andreistan26 
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>
Approved-by: Stefan Jumarea <stefanjumarea02@gmail.com>

**Pull request merged**